vpr/src/route/rr_graph_storage.h

```diff
@@ -0,0 +1,578 @@
+    void set_node_cost_index(RRNodeId, size_t new_cost_index);
+    void set_node_rc_index(RRNodeId, short new_rc_index);
+    void set_node_capacity(RRNodeId, short new_capacity);
+    void set_node_direction(RRNodeId, e_direction new_direction);
+    void set_node_side(RRNodeId, e_side new_side);
+
+    // This prefetechs hot RR node data required for optimization.
+    //
+    // Note: This is optional, but may lower time spent on memory stalls in
+    // some circumstances.
+    inline void prefetch_node(RRNodeId id) const {
+        VTR_PREFETCH(&storage_[id], 0, 0);
+    }
+
+    /****************
+     * Edge methods *
+     ****************/
+
+    // Edge initialization ordering:
+    //  1. Use reserve_edges/emplace_back_edge/alloc_and_load_edges to
+    //     initialize edges.  All edges must be added prior to calling any
+    //     methods that read edge data.
+    //
+    //     Note: Either arch_switch_inf indicies or rr_switch_inf should be
+    //     used with emplace_back_edge and alloc_and_load_edges.  Do not mix
+    //     indicies, otherwise things will be break.
+    //
+    //  2. The following methods read from the edge data, and lock out the
+    //     edge mutation methods (e.g. emplace_back_edge/alloc_and_load_edges):
+    //       - init_fan_in
+    //       - partition_edges
+    //       - count_rr_switches
+    //       - remap_rr_node_switch_indices
+    //       - mark_edges_as_rr_switch_ids
+    //
+    //  3. If edge_switch values are arch_switch_inf indicies,
+    //     remap_rr_node_switch_indices must be called prior to calling
+    //     partition_edges.
+    //
+    //     If edge_switch values are rr_switch_inf indices,
+    //     mark_edges_as_rr_switch_ids must be called prior to calling
+    //     partition_edges.
+    //
+    //  4. init_fan_in can be invoked any time after edges have been
+    //     initialized.
+    //
+    //  5. The following methods must only be called after partition_edges
+    //     have been invoked:
+    //      - edges
+    //      - configurable_edges
+    //      - non_configurable_edges
+    //      - num_edges
+    //      - num_configurable_edges
+    //      - edge_id
+    //      - edge_sink_node
+    //      - edge_switch
+
+    /* Edge mutators */
+
+    // Reserve at least num_edges in the edge backing arrays.
+    void reserve_edges(size_t num_edges);
+
+    // Adds ones edge.  This method is efficient if reserve_edges was called with
+    // the number of edges present in the graph.  This method is still
+    // amortized O(1), like std::vector::emplace_back, but both runtime and
+    // peak memory usage will be higher if reallocation is required.
+    void emplace_back_edge(RRNodeId src, RRNodeId dest, short edge_switch);
+
+    // Adds a batch of edges.
+    void alloc_and_load_edges(const t_rr_edge_info_set* rr_edges_to_create);
+
+    /* Edge finalization methods */
+
+    // Counts the number of rr switches needed based on fan in.
```

Comment:
to support mux size dependent switch delays.

Reply:
done